src/ng/rootScope.js

```diff
@@ -1181,10 +1181,12 @@ function $RootScopeProvider() {
 
         var self = this;
         return function() {
-          var indexOfListener = namedListeners.indexOf(listener);
-          if (indexOfListener !== -1) {
-            namedListeners[indexOfListener] = null;
+          var index = arrayRemove(namedListeners, listener);
+          if (index >= 0) {
             decrementListenerCount(self, 1, name);
+            if (index <= namedListeners.$$index) {
+              namedListeners.$$index--;
+            }
```

Comment:
A comment here explaining the logic will help in the future. E.g. "we are removing a listener while iterating over the list of listeners. Update the current $index if necessary to ensure no listener is skipped."